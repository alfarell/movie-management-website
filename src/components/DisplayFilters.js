import React, { useState, useEffect } from 'react';
import Axios from 'axios';
import { Menu } from 'antd';

const DisplayFilters = () => {
    const [genreList, setGenreList] = useState([]);
    const [selectedGenre, setSelectedGenre] = useState('Select genre')

    useEffect(() => {
        Axios.get(`${process.env.REACT_APP_BASE_URL}/genre/list?api_key=${process.env.REACT_APP_BASE_API_KEY}`)
            .then(res => {
                setGenreList(res.data.genres);
            })
            .catch(err => {

            });
    }, []);

    return (
        <div style={{ flex: 1, width: '100%', display: 'flex' }}>
            <Menu mode='horizontal'>
                <Menu.SubMenu title={<span>{selectedGenre}</span>} disabled={genreList.length === 0 ? true : false} >
                    {genreList.map(genre => {
                        return (
                            <Menu.Item key={genre.id} onClick={() => setSelectedGenre('Genre : ' + genre.name)}>{genre.name}</Menu.Item>
                        )
                    })}
                </Menu.SubMenu>
            </Menu>
        </div>
    );
};

export default DisplayFilters;
