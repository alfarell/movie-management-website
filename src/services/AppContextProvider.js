import React, { createContext, useState, useEffect } from 'react';
import _ from 'loadsh';
import Axios from 'axios';


export const MovieContext = createContext();
export const UserContext = createContext();

const AppContextProvider = ({ children }) => {
    const [movieList, setMovieList] = useState([]);
    const [genreList, setGenreList] = useState([]);
    const [selectedGenre, setSelectedGenre] = useState(null);
    const [sortOption, setSortOption] = useState('popularity.desc')
    const [pagination, setPagination] = useState(1);


    useEffect(() => {
        Axios.get(`${process.env.REACT_APP_BASE_URL}/genre/list?api_key=${process.env.REACT_APP_BASE_API_KEY}&language=en-US`)
            .then(res => {
                setGenreList(res.data.genres);
            })
            .catch(err => {
                console.log(err);
            });
    }, [genreList]);

    useEffect(() => {
        const discovers = {
            genre: selectedGenre ? selectedGenre.id : '',
            sort: sortOption,
        };

        setPagination(1);
        fetchMovie(discovers);
    }, [selectedGenre, sortOption]);

    const fetchMovie = ({ genre, sort }) => {
        Axios.get(`${process.env.REACT_APP_BASE_URL}/discover/movie?api_key=${process.env.REACT_APP_BASE_API_KEY}&language=en-US&include_adult=false&include_video=false&page=${pagination}&with_genres=${genre}&sort_by=${sort}`)
            .then(res => {
                setMovieList([...res.data.results])
            })
            .catch(err => {
                console.log(err);
            });
    }

    const handleLoadMore = () => {
        setPagination(pagination + 1);

        const genre = selectedGenre ? selectedGenre.id : '';
        const sort = sortOption;
        let page = pagination + 1

        Axios.get(`${process.env.REACT_APP_BASE_URL}/discover/movie?api_key=${process.env.REACT_APP_BASE_API_KEY}&language=en-US&include_adult=false&include_video=false&page=${page}&with_genres=${genre}&sort_by=${sort}`)
            .then(res => {
                setMovieList([...movieList, ...res.data.results])
            })
            .catch(err => {
                console.log(err);
            });
    }

    const addFavoriteMovie = (favoritedMovie) => {
        let currentMovie = JSON.parse(localStorage.getItem('favorited'));
        let addFavorited;

        if (currentMovie) {
            let check = _.find(currentMovie, favoritedMovie);
            if (check) {
                addFavorited = currentMovie.filter(movie => {
                    return movie.id !== favoritedMovie.id
                });
            } else {
                addFavorited = [...currentMovie, favoritedMovie]
            }
        } else {
            addFavorited = [favoritedMovie];
        }

        localStorage.setItem('favorited', JSON.stringify(addFavorited));
    }

    return (
        <MovieContext.Provider value={{
            movieList,
            setMovieList,
            genreList,
            setGenreList,
            selectedGenre,
            setSelectedGenre,
            sortOption,
            setSortOption,
            pagination,
            setPagination,
            handleLoadMore,
            addFavoriteMovie
        }}>
            <UserContext.Provider value={{ test: 'test' }}>
                {children}
            </UserContext.Provider>
        </MovieContext.Provider>
    );
};

export default AppContextProvider;
