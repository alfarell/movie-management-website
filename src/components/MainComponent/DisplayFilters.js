import React, { useContext } from 'react';
import { Menu, Tag } from 'antd';
import { MovieContext } from '../../services/AppContextProvider';

const DisplayFilters = () => {
    const { genreList, selectedGenre, setSelectedGenre, setSortOption } = useContext(MovieContext);

    return (
        <Menu mode='inline'>
            <div>
                {
                    selectedGenre
                        ? <Tag color='blue' closable onClose={() => setSelectedGenre({})} data-testid='selected-genre'>{selectedGenre.name}</Tag>
                        : null
                }
            </div>
            <Menu.SubMenu title='Sort By'>
                <Menu.Item onClick={() => setSortOption('popularity.desc')}>Popularity</Menu.Item>
                <Menu.Item onClick={() => setSortOption('release_date.desc')}>Latest Release</Menu.Item>
                <Menu.Item onClick={() => setSortOption('release_date.asc')}>Oldest Release</Menu.Item>
            </Menu.SubMenu>
            <Menu.SubMenu title={<span>Select Genre</span>} disabled={genreList.length === 0 ? true : false} >
                {genreList.map(genre => {
                    return (
                        <Menu.Item
                            key={genre.id}
                            onClick={() => setSelectedGenre({ id: genre.id, name: genre.name })}>
                            {genre.name}
                        </Menu.Item>
                    )
                })}
            </Menu.SubMenu>
        </Menu>
    );
};

export default DisplayFilters;
