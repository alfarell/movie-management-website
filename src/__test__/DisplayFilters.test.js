import React, { useContext } from 'react';
import renderer from 'react-test-renderer';
import AppContextProvider, { MovieContext } from '../services/AppContextProvider';
import DisplayFilters from '../components/DisplayFilters';
import { render, fireEvent } from '@testing-library/react';
import { mockComponent } from 'react-dom/test-utils';

describe('Display Filter Snapshot', () => {
    it('render display filters', () => {
        const component = renderer.create(
            <AppContextProvider>
                <DisplayFilters />
            </AppContextProvider>
        );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });
});

describe('Display Filters functions and render genres', () => {
    const TestComponent = () => {
        const { sortOption } = useContext(MovieContext);

        return (
            <div>
                <p data-testid='sort-by'>{sortOption}</p>
            </div>
        );
    };

    it('should set sort option to popularity', () => {
        const component = render(
            <AppContextProvider>
                <TestComponent />
                <DisplayFilters />
            </AppContextProvider>
        );

        fireEvent.click(component.getByText('Sort By'), onclick);
        fireEvent.click(component.getByText('Popularity'), onclick);

        expect(component.getByTestId('sort-by')).toHaveTextContent('popularity.desc');
    });

    it('should set sort option to latest release', () => {
        const component = render(
            <AppContextProvider>
                <TestComponent />
                <DisplayFilters />
            </AppContextProvider>
        );

        fireEvent.click(component.getByText('Sort By'), onclick);
        fireEvent.click(component.getByText('Latest Release'), onclick);

        expect(component.getByTestId('sort-by')).toHaveTextContent('release_date.desc');
    });

    it('should set sort option to oldest release', async () => {
        const component = render(
            <AppContextProvider>
                <TestComponent />
                <DisplayFilters />
            </AppContextProvider>
        );

        fireEvent.click(component.getByText('Sort By'), onclick);
        fireEvent.click(component.getByText('Oldest Release'), onclick);

        expect(component.getByTestId('sort-by')).toHaveTextContent('release_date.asc');
    });

    it('should render the genres', () => {
        const genreList = [
            {
                id: 1,
                name: 'Test'
            }
        ];

        const component = render(
            <MovieContext.Provider value={{ genreList }}>
                <DisplayFilters />
            </MovieContext.Provider>
        );
        expect(component).toMatchSnapshot();

        fireEvent.click(component.getByText('Select Genre'), onclick);
        expect(component.getByText('Test')).toBeTruthy();
    });
});
