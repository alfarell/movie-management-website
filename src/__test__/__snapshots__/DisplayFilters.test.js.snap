// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Filter Snapshot render display filters 1`] = `
<ul
  className="ant-menu ant-menu-light ant-menu-root ant-menu-inline"
  onMouseEnter={[Function]}
  onTransitionEnd={[Function]}
  role="menu"
  style={Object {}}
>
  <div />
  <li
    className="ant-menu-submenu ant-menu-submenu-inline"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="menuitem"
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      className="ant-menu-submenu-title"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="button"
      style={
        Object {
          "paddingLeft": 24,
        }
      }
      title="Sort By"
    >
      Sort By
      <i
        className="ant-menu-submenu-arrow"
      />
    </div>
    <div />
  </li>
  <li
    className="ant-menu-submenu ant-menu-submenu-inline ant-menu-submenu-disabled"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="menuitem"
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      className="ant-menu-submenu-title"
      role="button"
      style={
        Object {
          "paddingLeft": 24,
        }
      }
    >
      <span>
        Select Genre
      </span>
      <i
        className="ant-menu-submenu-arrow"
      />
    </div>
    <div />
  </li>
</ul>
`;

exports[`Display Filters functions and render genres should render the genres 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="ant-menu ant-menu-light ant-menu-root ant-menu-inline"
        role="menu"
      >
        <div />
        <li
          class="ant-menu-submenu ant-menu-submenu-inline"
          role="menuitem"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
            style="padding-left: 24px;"
            title="Sort By"
          >
            Sort By
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
          <div />
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-inline"
          role="menuitem"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
            style="padding-left: 24px;"
          >
            <span>
              Select Genre
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
          <div />
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="ant-menu ant-menu-light ant-menu-root ant-menu-inline"
      role="menu"
    >
      <div />
      <li
        class="ant-menu-submenu ant-menu-submenu-inline"
        role="menuitem"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="ant-menu-submenu-title"
          role="button"
          style="padding-left: 24px;"
          title="Sort By"
        >
          Sort By
          <i
            class="ant-menu-submenu-arrow"
          />
        </div>
        <div />
      </li>
      <li
        class="ant-menu-submenu ant-menu-submenu-inline"
        role="menuitem"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="ant-menu-submenu-title"
          role="button"
          style="padding-left: 24px;"
        >
          <span>
            Select Genre
          </span>
          <i
            class="ant-menu-submenu-arrow"
          />
        </div>
        <div />
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
